### PR TITLE
Textview, a simple text viewer for big files (e.g., Qemu logs)

### PR DESCRIPTION
**Pull request by laerreal**
Minimum features set for now.

## v2:
- fixup coding style and typos
- add docstring for `OffsetsStream.__init__`
- use `divmod` in `offset2line_offset`
- extend `SoftStream` docstring
- fixup scrolling by mouse wheel under Windows
- OS independent keyboard bindings based on `KeyboardSettings`
- fixup `OffsetsStream` doctest about `0xDEADBEEF` under Windows Py2
- `TextCanvas`: utf-8 encoding with "replace" error handling
   (fixes troubles with `\` in raw data).

## v3:
- fixup coding stile and typos
- `wrap_wheel_handler` tested on Windows 10
- `HKCombobox` use OS-independent `COPY_KEYCODE` too

### CODE REVIEW

File: common/os_wrappers.py [hunk 1]
```diff
@@ -89,30 +89,36 @@ def cli_repr(obj):
...
 else:
-    def bind_all_mouse_wheel(*_, **__):
-        from sys import ( # required only there
-            stderr
-        )
-        stderr.write("bind_all_mouse_wheel is not implemented for %s\n" % OS)
+    def wrap_wheel_handler(*__):
+        return tuple()
+
+    from sys import ( # required only there
```
Comment (BDanAnd) [inline common/os_wrappers.py]: "here"?

File: common/line_index.py [hunk 2]
```diff
@@ -0,0 +1,178 @@
+__all__ = [
+    "re_newline",
+    "LineIndex",
+    "LineIndexIsNotReady"
+]
```
Comment (BDanAnd) [inline common/line_index.py]:
```
__all__ = [
    "re_newline"
  , "LineIndex"
  , "LineIndexIsNotReady"
]
```

File: common/line_index.py [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+
+
```
Comment (BDanAnd) [inline common/line_index.py]: One or two newlines before the first import?

File: common/soft_stream.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+    def __init__(self, size):
+        self._offset = 0
+        self._size = size
+
+    def tell(self):
+        return self._offset
+
+    def seek(self, offset, whence = 0):
+        # TODO: keep resulting offset within [0, self._size)
+        return getattr(self, "_seek_%d" % whence)(offset)
```
Comment (BDanAnd) [inline common/soft_stream.py:21]: Is `getattr` better than an `if-else` block?

File: common/offsets_stream.py [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+>>> s.read() == (b"0000" + b"0004" + b"0008" + b"000c" + b"0010" + b"0014")
+True
+>>> s.read(1) == b''
+True
+>>> s.read() == b''
+True
+    """
+
+    def __init__(self,
+        word_size_bits = 4, # 16 byte ASCII word contains its 8 byte offset
```
Comment (BDanAnd) [inline common/offsets_stream.py]: Maybe the word "HEX" is needed here to clarify the idea. Maybe "represents".

File: common/offsets_stream.py [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+>>> s.read(1) == b'0'
+True
+>>> s.read(7) == b'0000000'
+True
+>>> s.tell()
+8
+>>> s.read(8) == b'00000008'
+True
+>>> s.seek(0xDEADBEEF)
+3735928559
```
Comment (BDanAnd) [inline common/offsets_stream.py]: On Python 2.7.13 (Windows):
```
File "...\qdt\common\offsets_stream.py", line 25, in common.offsets_stream.OffsetsStream
Failed example:
    s.seek(0xDEADBEEF)
Expected:
    3735928559
Got:
    3735928559L
```

File: common/line_no_stream.py [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+>>> offset2line_offset(lines_data(0, 1000).find(b"321"))
+(321, 0)
+>>> offset2line_offset(lines_data(0, 10000).find(b"4321"))
+(4321, 0)
+>>> offset2line_offset(lines_data(0, 10000).find(b"\\r\\n321\\r\\n"))
+(320, 3)
+    """
+    decade = offset2decade(offset, eol_size = eol_size)
+
+    # Note, required cache aall_decade_offsets[eol_size] is already defined by
```
Comment (BDanAnd) [inline common/line_no_stream.py]: `all_decade_offsets`

File: common/line_no_stream.py [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+    # `offset2decade` above.
+    decade_intra_offset = offset - all_decade_offsets[eol_size][decade - 1]
+
+    line_size = decade + eol_size
+    lineno = decade_intra_offset // line_size
+    line_offset = decade_intra_offset % line_size
```
Comment (BDanAnd) [inline common/line_no_stream.py]: Maybe:
```
lineno, line_offset = divmod(decade_intra_offset, line_size)
```

File: widgets/text_canvas.py [hunk 9]
```diff
@@ -0,0 +1,348 @@
...
+)
+from common import (
+    bind_mouse_wheel,
+    LineIndex,
+)
+
+
+class TextCanvas(Canvas, object):
+    """ Shows big raw stream with text.
+    """
```
Comment (BDanAnd) [inline widgets/text_canvas.py]: Single " for one line help text (by coding style).

File: widgets/text_canvas.py [hunk 10]
```diff
@@ -0,0 +1,348 @@
...
+
+        # TODO: make it configurable
+        fonts = [
+            Font(font = ("Courier", 10, NORMAL)),
+            Font(font = ("Courier", 10, BOLD)),
+        ]
+        self._main_font, self._lineno_font = fonts
+        self._linespace = max(f.metrics("linespace") for f in fonts)
+        self._ylinepadding = 1
+        self._lineno_pading = 10
```
Comment (BDanAnd) [inline widgets/text_canvas.py]: `padding` and `pading`...
How many letters `d`?

File: common/os_wrappers.py [hunk 11]
```diff
@@ -89,30 +89,36 @@ def cli_repr(obj):
...
-        w.bind_all("<ButtonPress-5>", scroll_down, add = add)
+        yield scroll_down
 
 elif OS == "Windows":
-    def bind_all_mouse_wheel(w, handler, add = None):
-
+    def wrap_wheel_handler(handler):
         def scale(e):
+            "<MouseWheel>"
             e.delta //= -120
```
Comment (BDanAnd) [inline common/os_wrappers.py]: Scrolling inverted in `text viewer` on Win10 (wheel up - text down).

File: widgets/text_canvas.py [hunk 12]
```diff
@@ -0,0 +1,348 @@
...
+        self.x_offset = int(offset_f * float(self._max_text_width))
+
+    def _xview_scroll(self, step, unit):
+        step_i = int(step)
+        if unit == "pages":
+            shift = step_i * self._page_width
+        elif unit == "units": # canvas coordinates
+            shift = step_i
+        else:
+            raise ValueError("Unsupported scsroll unit: " + unit)
```
Comment (BDanAnd) [inline widgets/text_canvas.py]: `scroll`

File: widgets/text_canvas.py [hunk 13]
```diff
@@ -0,0 +1,348 @@
...
+        self.lineno = int(lineno_f)
+
+    def _yview_scroll(self, step, unit):
+        step_i = int(step)
+        if unit == "pages":
+            shift = step_i * self._page_size
+        elif unit == "units": # lines
+            shift = step_i
+        else:
+            raise ValueError("Unsupported scsroll unit: " + unit)
```
Comment (BDanAnd) [inline widgets/text_canvas.py]: `scroll`

File: misc/textview.py [hunk 14]
```diff
@@ -0,0 +1,87 @@
+from argparse import (
```
Comment (BDanAnd) [inline misc/textview.py:1]: `horizontal` in commit description.

File: misc/textview.py [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+        row = 0
+        self.rowconfigure(row, weight = 1)
+
+        self._text = text = TextCanvas(self)
+        text.grid(row = row, column = 0, sticky = "NESW")
+
+        add_scrollbars_native(self, text, row = row)
+        row += 1;
+
+        row += 1; self.rowconfigure(row, weight = 0)
```
Comment (BDanAnd) [inline misc/textview.py]: Really need `;`?

File: widgets/text_canvas.py [hunk 16]
```diff
@@ -360,3 +443,103 @@ def _on_prior(self, __):
...
+            rel_lineidx -= 1
+
+        lineidx = self.lineidx + rel_lineidx
+
+        # get char index
+        try:
+            line = picked_lines[rel_lineidx]
+        except IndexError:
+            # _lineidx_charidx is only used for x, y of mouse pointer.
+            # So, where are two cases:
```
Comment (BDanAnd) [inline widgets/text_canvas.py]: there

File: widgets/text_canvas.py [hunk 17]
```diff
@@ -543,3 +595,11 @@ def _on_motion(self, e):
                 self._sel_start = first
                 self._sel_limit = (lineidx, charidx + 1)
             self.draw()
+
+    def _on_ctrl_key(self, e):
+        if e.keycode == 54: # Ctrl-C: copy selected
```
Comment (BDanAnd) [inline widgets/text_canvas.py]: Doesn't work on Windows (keycode 67 for C).

File: widgets/text_canvas.py [hunk 18]
```diff
@@ -543,3 +595,11 @@ def _on_motion(self, e):
...
+            blob = self.selected_blob
+            text = blob.decode(self._encoding)
```
Comment (BDanAnd) [inline widgets/text_canvas.py]: When the selected text ends by `\`:
```
Exception in Tkinter callback
Traceback (most recent call last):
  File "...\Python37\lib\tkinter\__init__.py", line 1705, in __call__
    return self.func(*args)
  File "...\qdt\widgets\text_canvas.py", line 603, in _on_ctrl_key
    text = blob.decode(self._encoding)
UnicodeDecodeError: 'unicodeescape' codec can't decode byte 0x5c in position 13: \ at end of string
```

File: common/line_index.py (hunk 3 above)
Comment (laerreal) [reply common/line_index.py]: It's <strike>not</strike> specified currently. <strike>At least</strike> one.

File: common/soft_stream.py (hunk 4 above)
Comment (laerreal) [reply common/soft_stream.py:21]: This allow to keep different `seek` logic in dedicated methods. Handy to override. Shorter.

File: common/line_no_stream.py (hunk 8 above)
Comment (laerreal) [reply common/line_no_stream.py]: Wow, wow

File: common/soft_stream.py [hunk 19]
```diff
@@ -0,0 +1,50 @@
+__all__ = [
+    "SoftStream"
+]
+
+
+class SoftStream(object):
+    """ A base class for implementing software generated byte streams.
```
Comment (BDanAnd) [inline common/soft_stream.py:7]: Single " for one line help text (by coding style).

File: misc/textview.py (hunk 14 above)
Comment (laerreal) [reply misc/textview.py:1]: This error now is a history now =)

Comment (laerreal) [reply misc/textview.py:1]: We cannot change name of what is gone forever...

File: common/os_wrappers.py [hunk 20]
```diff
@@ -106,7 +106,9 @@ def scroll_down(e):
     def wrap_wheel_handler(handler):
         def scale(e):
             "<MouseWheel>"
-            e.delta //= -120
+            # This divisor is tested on:
+            # - Windows 7 SP 1 x64
```
Comment (BDanAnd) [inline common/os_wrappers.py:111]: Please add Windows 10 1709 x64

File: widgets/text_canvas.py [hunk 21]
```diff
@@ -0,0 +1,348 @@
+__all__ = [
+    "TextCanvas"
+]
+
+
```
Comment (BDanAnd) [inline widgets/text_canvas.py]: One newline.

File: common/line_no_stream.py [hunk 22]
```diff
@@ -0,0 +1,154 @@
+__all__ = [
+    "LineNoStream"
+]
+
+
```
Comment (BDanAnd) [inline common/line_no_stream.py]: One newline.

File: common/offsets_stream.py [hunk 23]
```diff
@@ -0,0 +1,99 @@
+__all__ = [
+    "OffsetsStream"
+]
+
+
```
Comment (BDanAnd) [inline common/offsets_stream.py]: One newline.

File: widgets/text_canvas.py (hunk 10 above)
Comment (BDanAnd) [reply widgets/text_canvas.py]: `padding` and `pading`...
How many letters `d`?

File: widgets/text_canvas.py (hunk 12 above)
Comment (BDanAnd) [reply widgets/text_canvas.py]: `scroll`

File: widgets/text_canvas.py (hunk 13 above)
Comment (BDanAnd) [reply widgets/text_canvas.py]: `scroll`

File: widgets/text_canvas.py (hunk 9 above)
Comment (BDanAnd) [reply widgets/text_canvas.py]: Single " for one line help text (by coding style).

File: widgets/hotkey.py [hunk 24]
```diff
@@ -30,6 +32,44 @@
...
+class CurrentKeyboard:
+    "A back-end to implement OS & locale independent keyboard key bindings."
+
+    @lazy
+    def _current_os_keycodes(self):
+        with KeyboardSettings() as kbd:
+            try:
+                codes = kbd.os_codes[os_name]
+            except KeyError:
+                print("No keyboard layout for OS %s, using posix " % os_name)
```
Comment (BDanAnd) [inline widgets/hotkey.py]: `using posix"`

File: widgets/hotkey.py [hunk 25]
```diff
@@ -70,17 +70,18 @@ def SELECT_ALL_KEYCODE(self):
         return self.get_keycode(3, 1)
 
 
-class HKGeneric:
+class HKGeneric(CurrentKeyboard):
 
     def _on_ctrl_key_generic(self, event):
```
Comment (BDanAnd) [inline widgets/hotkey.py:75]: `HKCombobox._on_ctrl_key` need fix too.

File: misc/textview.py (hunk 14 above)
Comment (BDanAnd) [reply misc/textview.py:1]: You can edit the description during the rebase :(

Comment (laerreal) [reply misc/textview.py:1]: Yes, but patch with correct message never existed.

**Comment by laerreal** [comment]
v3 is ready

**Pull request merged**